man/man1/asc2clm.1
.TH asc2clm 1  "USER COMMANDS"
.SH NAME
asc2clm \- convert ARC/Info ASCII files to clm data files for LPJmL
.SH SYNOPSIS
.B asc2clm
[\-h] [\-f] [\-firstyear y] [\-grid file] [\-nbands n] [\-{int|float}] [-version v] [\-scale s] [\-json]
.I ascfile ... clmfile
.SH DESCRIPTION
Program converts ARC/Info ASCII (plain text) input data files into LPJ climate data file.
.SH OPTIONS
.TP
\-h,\--help
Print a short help text.
.TP
\-f
Force overwrite of existing CLM file.
.TP
\-firstyear y
Set first simulation year in CLM file, default is 1901.
.TP
\-grid file
Write LPJ grid file.
.TP
\-nbands n
Set number of bands of CLM file, default is 12.
.TP
\-header s
Set header string of CLM file, default is LPJCLIM.
.TP
\-int
Write integer data into CLM file.
.TP
\-float
Write float data into CLM file.
.TP
\-version v
Set version of CLM file, default is 3.
.TP
\-scale s
Scale data by a factor.
.TP
\-json
JSON metafile is created with suffix '.json'.
.TP
.I ascfile
Filename(s) of ARC/info ASCII data.
.TP
.I clmfile
Filename of clm data file written.
.SH EXAMPLES
.TP
Create clm data and grid file from ARC/Info dataset and scale data by 10:
.B asc2clm
-grid grid.clm -scale 10 tmp_1901_*.asc tmp_1901.clm
.PP
.SH EXIT STATUS
.B asc2clm
returns a zero exit status if the clm file is created.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
lpjml(1), txt2clm(1), cru2clm(1), grid2clm(1), cft2clm(1), arr2clm(1), catclm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]


### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: David Hötten (davidho@pik-potsdam.de)

### Added

- Warning message added in `lpjml`, `lpjcheck`,`cdf2soil`, `txt2grid` and `cdf2coord` if cell size cannot be resolved by grid data in short format.
- Function `isfloatcoord()` added in `coord.c` to check whether coordinate must be stored as a float variable, test function `test_isfloatcoord.c` added.
- Test functions for `vmaxlimit_tree`, `vmaxlimit_grass`, `vmaxlimit_crop`, `bisect`, `leftmostzero` and `getbufferavg` functions added.

### Changed

- Error about first simulation year is less than first year of climate data replaced by warning in `initclimate.c`.
- First and last simulation year in spinup changed from 1901 to 1840 in `lpjml_config.cjson`.
- Number of spinup years changed from 3500 to 3439 in `lpjml_config.cjson` in order to get bit-identical results.
- Test files for unit test added to tar and zip file in `Makefile`.

### Removed

- Unused source file `src/lpj/part.c` removed.

### Fixed

- Expected number of bands set correctly for raw input files in `openinputdata.c`.
- Hard-coded filename `test.lpj` replaced by temporary filename in `test_append.c`, `test_getnoread.c`, `test_indexarray.c`, `test_isdefined.c`, `test_isnull.c`, `test_restart.c`, and `test_writename.c` in order to allow parallel execution of tests.
- Arguments corrected in calls of lpjml functions in `lpj_poem.c`.
- Missing `-print_noread` option added in `INSTALL`.
- `setupannual_grid.c` corrected to compile with `-DIMAGE -DCOUPLED` flag.
- Error message for different datatypes corrected in `lpcat.c`.
- Variables initialized to zero and NULL in `phenology_gsi.c`, `lpjcat.c` and `restart2yaml.c` to avoid warnings in `gcc`.
- Man page for `json2restart` corrected.
- Missing option `-nooutput` added to man page of `lpjfiles`.
- Help text for `-h` option of `printclm` corrected.
- Check for NULL pointer for `pfts` array corrected in `lpjcat.c`.
- Double entry of `getlag.$O` removed in `src/soil/Makefile`.


## [5.10.1] - 2025-09-30

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de)
- code review: Maik Billing (billing@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)

### Added

- Variable `phen` added to output in `fprintpft.c`.
- Compile option `NRECOVERY_COST` added to `INSTALL` and man page of `configure.sh`.
- Check for valid `swc_bnf` parameters added in `fscanpftpar.c`.
- Wirth et al. (2024)  article added to `REFERENCES`.
- New parameter `"tscal_b"` added in `lpjparam.cjson`.
- New virtual function `getb()` added to type `Pft`.

### Changed

- 2 boolean flags `phen_to_one` and `rainyseason` added to tree parameter. These are used in `phenology_gsi.c` instead of string compare with tree names.
- Unused argument `aet_layer` removed from `nitrogen_stress()` function.
- `phi_bnf` parameter calculated from `swc_bnf` parameter instead of reading from `pft.cjson` file in `fscanpftpar.c`.
- Hard-coded parameter in temperature dependent calculation of `b` replaced by global parameter `tscal_b`.
- Calculation of `b` in `setupannual_grid.c` replaced by call to new virtual PFT function `getb()`.

### Removed

- Parameter `phi_bnf` removed from `pft.cjson`.
- Variable `b` removed from restart file.
- Unused macro `RESTART_VERSION` removed from `header.h`.

### Fixed

- Arguments of `npp` function corrected in `daily_woodplantation.c`. The current version leads to a too strong reduction in NPP.
- `nuptake_crop.c` changed to compile with `-DDEBUG_N` option.
- `update_daily_cell.c` changed to compile with `-DDEBUG` option.
- `daily_agriculture_tree.c` changed to compile with `-DDEBUG2` option.
- Typo in error message in `bstruct_writearrayindex.c` corrected.
- Error message corrected in `bstruct_wopen.c`.
- Error number corrected in `bstruct_finish.c`.
- Argument `temp` added to `f_turnover_tree` function to compile with `-DNRECOVERY_COST` option.


## [5.10.0] - 2025-08-14

### Contributors

- author: Jens Heinke (heinke@pik-potsdam.de), Sibyll Schaphoff (sibylls@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Added

- option to specify year after which deforestation is no longer feeding product pools but biomass is burnt instead: `luc_burn_startyear` (disabled)
- option to specify costs for N recovery upon turnover (pre-compiler switch `NRECOVERY_COST`, disabled)

### Changed

- parameter changes:
  - `alphaa` from 0.5 to 0.7 for natural vegetation;
  - `theta` from 0.9 to 0.95;
  - `fn_turnover` PFT1: 0.5, PFT2: 0.5, PFT3: 0.7, PFT4: 0.3, PFT5: 0.3, PFT6: 0.7, PFT7: 0.3, PFT8: 0.3; updated values from [Sophia et al. 2024](https://doi.org/10.5194/bg-21-4169-2024) for all other PFTs except crops
  - `cnratio_leaf.low/median/high` for all PFTs updated from TRY data
  - `b` (ratio of dark respiration Rd to maximum photosynthetic capacity Vcmax) set to 0.031 for C3 and C4 plants (value for 25°C, from [Wang et al. 2020](https://doi.org/10.1111/gcb.14980), consistent with TRY data)

src/utils/asc2clm.c
/**************************************************************************************/
/**                                                                                \n**/
/**                      a  s  c  2  c  l  m  .  c                                 \n**/
/**                                                                                \n**/
/**     Program converts ARC/Info ASCII (plain text) input data into the           \n**/
/**     CLM format for LPJ                                                         \n**/
/**     Input must be in the following form:                                       \n**/
/**     ncols   288                                                                \n**/
/**     nrows   288                                                                \n**/
/**     xllcorner       98.021                                                     \n**/
/**     yllcorner       12.021                                                     \n**/
/**     cellsize        0.042                                                      \n**/
/**     NODATA_value    -9999.000                                                  \n**/
/**     3636.0 3737.0 ...                                                          \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"
#include <sys/stat.h>

#define USAGE "Usage: asc2clm [-h] [-f] [-firstyear y] [-grid file] [-nbands n] [-nstep n] [-header s]\n       [-version v] [-{int|float}] [-scale s] [-json] infile ... clmfile\n"
#define ERR_USAGE USAGE "\nTry \"asc2clm --help\" for more information.\n"
#define FIRSTYEAR 1901

int main(int argc,char **argv)
{
  FILE *file,*gridfile,*out;
  Bool first,isjson;
  float scale;
  int row,col,nrows,ncols,nyear,nbands,nodata_int,nstep,version;
  int nrows_first,ncols_first,ncell_first;
  float xllcorner_first,yllcorner_first,cellsize_first;
  float xllcorner,yllcorner,cellsize,nodata,value;
  short **data=NULL;
  int **data_int=NULL,value_int;
  float **data_float=NULL;
  char *endptr,*arglist,*out_json;
  Filename coord_filename;
  Type coord_type=LPJ_FLOAT;
  Coord grid;
  Type type;
  Header header;
  char *head;
  int i,m,y;
  Bool force;
  struct stat buf;
  char c;
  header.firstyear=FIRSTYEAR;
  header.firstcell=0;
  header.datatype=LPJ_SHORT;
  nyear=1;
  nbands=NMONTH;
  nstep=1;
  header.order=CELLYEAR;
  scale=1;
  head=LPJ_CLIMATE_HEADER;
  gridfile=NULL;
  force=isjson=FALSE;
  type=LPJ_SHORT;
  version=LPJ_CLIMATE_VERSION;
  /* parse command line options */
  for(i=1;i<argc;i++)
    if(argv[i][0]=='-')
    {
      if(!strcmp(argv[i],"-h") || !strcmp(argv[i],"--help"))
      {
        printf("   asc2clm (" __DATE__ ") help\n"
               "   ==========================\n\n"
               "Convert gridded ASCII files to clm data files for LPJmL version %s\n\n"
               USAGE
               "\nArguments:\n"
               "-h,--help    print this help text\n"
               "-f           force overwrite of output file\n"
               "-firstyear y first year, default is %d\n"
               "-grid file   create grid file\n" 
               "-nbands n    number of bands, default is 12\n"
               "-nstep n     number of steps, default is 1\n"
               "-header s    header string, default is " LPJ_CLIMATE_HEADER "\n"
               "-version v   version of clm file, default is %d\n"
               "-scale s     scale data by a factor of s\n"
               "-int         write integer data into clm file\n"
               "-float       write float data into clm file\n"
               "infile       filename(s) of gridded data file\n"
               "clmfile      filename of clm data file\n\n"
                "(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file\n",
                getversion(),FIRSTYEAR,LPJ_CLIMATE_VERSION);
        return EXIT_SUCCESS;
      }
      else if(!strcmp(argv[i],"-firstyear"))
      {
        if(i==argc-1)
        {
          fputs("Argument missing after '-firstyear' option.\n"
                ERR_USAGE,stderr);
          return EXIT_FAILURE;
        }
        header.firstyear=strtol(argv[++i],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-firstyear'.\n",
                  argv[i]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-json"))
        isjson=TRUE;
      else if(!strcmp(argv[i],"-f"))
        force=TRUE;
      else if(!strcmp(argv[i],"-int"))
        type=LPJ_INT;
      else if(!strcmp(argv[i],"-float"))
        type=LPJ_FLOAT;
      else if(!strcmp(argv[i],"-grid"))
      {
        if(i==argc-1)
        {
          fputs("Argument missing after '-grid' option.\n"
                ERR_USAGE,stderr);
          return EXIT_FAILURE;
        }
        gridfile=fopen(argv[++i],"wb");
        if(gridfile==NULL)
        {
          fprintf(stderr,"Error creating '%s': %s.\n",argv[i],strerror(errno));
          return EXIT_FAILURE;
        }
        coord_filename.name=argv[i];
        coord_filename.fmt=CLM;
        fwriteheader(gridfile,&header,LPJGRID_HEADER,LPJGRID_VERSION);
      }
      else if(!strcmp(argv[i],"-nbands"))
      {
        if(i==argc-1)
        {
          fputs("Argument missing after '-nbands' option.\n"
                ERR_USAGE,stderr);
          return EXIT_FAILURE;
        }
        nbands=strtol(argv[++i],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-nbands'.\n",
                  argv[i]);
          return EXIT_FAILURE;
        }
        if(nbands<1)
        {
          fputs("Error: number of bands must be >0.\n",stderr);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-nstep"))
      {
        if(i==argc-1)
        {
          fputs("Argument missing after '-nstep' option.\n"
                ERR_USAGE,stderr);
          return EXIT_FAILURE;
        }
        nstep=strtol(argv[++i],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-nstep'.\n",
                  argv[i]);
          return EXIT_FAILURE;
        }
        if(nstep<1)
        {
          fputs("Error: number of steps must be >0.\n",stderr);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-version"))
      {
        if(i==argc-1)
        {
          fputs("Argument missing after '-version' option.\n"
                ERR_USAGE,stderr);
          return EXIT_FAILURE;
        }
        version=strtol(argv[++i],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-version'.\n",
                  argv[i]);
          return EXIT_FAILURE;
        }
        if(version<1)
        {
          fputs("Error: Version must be  >0.\n",stderr);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-header"))
      {
        if(i==argc-1)
        {
          fputs("Argument missing after '-header' option.\n"
                ERR_USAGE,stderr);
          return EXIT_FAILURE;
        }
        head=argv[++i];
      }
      else if(!strcmp(argv[i],"-scale"))
      {
        if(i==argc-1)
        {
          fputs("Argument missing after '-scale' option.\n"
                ERR_USAGE,stderr);
          return EXIT_FAILURE;
        }
        scale=(float)strtod(argv[++i],&endptr);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-scale'.\n",
                  argv[i]);
          return EXIT_FAILURE;
        }
        if(scale==0.0)
        {
          fputs("Error: scale factor must not be zero.\n",stderr);
          return EXIT_FAILURE;
        }
      }
      else
      {
        fprintf(stderr,"Error: invalid option '%s'.\n"
                ERR_USAGE,argv[i]);
        return EXIT_FAILURE;
      }
    }
    else
      break;
  if(argc<i+2)
  {
    fputs("Filename(s) missing.\n"
          ERR_USAGE,stderr);
    return EXIT_FAILURE;
  }
  argc-=i;
  argv+=i;
  if((argc-1)% (nbands*nstep))
  {
    fputs("Number of input files is not multiple of nbands.\n",stderr);
    return EXIT_FAILURE;
  }
  nyear=(argc-1)/nbands/nstep;
  if(!force)
  {
    if(!stat(argv[argc-1],&buf))
    {
      fprintf(stderr,"File '%s' already exists, overwrite (y/n)?\n",argv[argc-1]);
      scanf("%c",&c);
      if(c!='y')
        return EXIT_FAILURE;
    }
  }
  out=fopen(argv[argc-1],"wb");
  if(out==NULL)
  {
    fprintf(stderr,"Error creating '%s': %s.\n",argv[argc-1],strerror(errno));
    return EXIT_FAILURE;
  }
  header.nbands=nbands;
  header.nbands=nstep;
  fwriteheader(out,&header,head,LPJ_CLIMATE_VERSION);
  first=TRUE;
  nrows_first=0;
  ncols_first=0;
  xllcorner_first=0;
  yllcorner_first=0;
  cellsize_first=0;
  for(y=0;y<nyear;y++)
  {
    for(m=0;m<nbands*nstep;m++)
    {
      file=fopen(argv[y*nbands*nstep+m],"r");
      if(file==NULL)
      {
        fprintf(stderr,"Error opening '%s': %s.\n",argv[y*nbands*nstep+m],strerror(errno));
        return EXIT_FAILURE;
      }
      /* parse header  */
      if(fscanf(file,"ncols %d\n",&ncols)!=1)
      {
        fputs("Error reading ncols.\n",stderr);
        return EXIT_FAILURE;
      }
      if(fscanf(file,"nrows %d\n",&nrows)!=1)
      {
        fputs("Error reading nrows.\n",stderr);
        return EXIT_FAILURE;
      }
      if(fscanf(file,"xllcorner %f\n",&xllcorner)!=1)
      {
        fputs("Error reading xllcorner.\n",stderr);
        return EXIT_FAILURE;
      }
      if(fscanf(file,"yllcorner %f\n",&yllcorner)!=1)
      {
        fputs("Error reading yllcorner.\n",stderr);
        return EXIT_FAILURE;
      }
      if(fscanf(file,"cellsize %f\n",&cellsize)!=1)
      {
        fputs("Error reading cellsize.\n",stderr);
        return EXIT_FAILURE;
      }
      if(type==LPJ_INT)
      {
        if(fscanf(file,"NODATA_value %d\n",&nodata_int)!=1)
        {
          fputs("Error reading nodata value.\n",stderr);
          return EXIT_FAILURE;
        }
      }
      else
      {
        if(fscanf(file,"NODATA_value %f\n",&nodata)!=1)
        {
          fputs("Error reading nodata value.\n",stderr);
          return EXIT_FAILURE;
        }
      }
      if(first)
      {
        switch(type)
        {
          case LPJ_INT:
            data_int=newmatrix(int,nrows*ncols,nbands*nstep);
            if(data_int==NULL)
            {
              fputs("Error allocating memory for data.\n",stderr);
              return EXIT_FAILURE;
            }
            break;
          case LPJ_SHORT:
            data=newmatrix(short,nrows*ncols,nbands*nstep);
            if(data==NULL)
            {
              fputs("Error allocating memory for data.\n",stderr);
              return EXIT_FAILURE;
            }
            break;
          case LPJ_FLOAT:
            data_float=newmatrix(float,nrows*ncols,nbands*nstep);
            if(data_float==NULL)
            {
              fputs("Error allocating memory for data.\n",stderr);
              return EXIT_FAILURE;
            }
            break;
          default:
            fprintf(stderr,"Invalid datatype %d in '%s'.\n",type,argv[y*nbands*nstep+m]);
            return EXIT_FAILURE;
        } /* of 'switch' */
        nrows_first=nrows;
        ncols_first=ncols;
        xllcorner_first=xllcorner;
        yllcorner_first=yllcorner;
        cellsize_first=cellsize;
      }
      else if(nrows_first!=nrows)
      {
        fprintf(stderr,"Number of rows %d in '%s' differs from %d.\n",
                nrows,argv[y],nrows_first);
        return EXIT_FAILURE;
      }
      else if(ncols_first!=ncols)
      {
        fprintf(stderr,"Number of columns %d in '%s' differs from %d.\n",
                ncols,argv[y],ncols_first);
        return EXIT_FAILURE;
      }
      else if(xllcorner_first!=xllcorner)
      {
        fprintf(stderr,"xllcorner %g in '%s' differs from %g.\n",
                xllcorner,argv[y],xllcorner_first);
        return EXIT_FAILURE;
      }
      else if(yllcorner_first!=yllcorner)
      {
        fprintf(stderr,"yllcorner %g in '%s' differs from %g.\n",
                yllcorner,argv[y],yllcorner_first);
        return EXIT_FAILURE;
      }
      else if(cellsize_first!=cellsize)
      {
        fprintf(stderr,"cellsize %g in '%s' differs from %g.\n",
                cellsize,argv[y],cellsize_first);
        return EXIT_FAILURE;
      }
      header.cellsize_lon=header.cellsize_lat=cellsize;
      header.ncell=0;
      if(type==LPJ_INT)
        for(row=0;row<nrows;row++)
          for(col=0;col<ncols;col++)
          {
            if(fscanf(file,"%d",&value_int)!=1)
            {
              fprintf(stderr,"Error reading data at (%d,%d) at year %d and band %d.\n",
                      row+1,col+1,header.firstyear+y,m+1);
              return EXIT_FAILURE;
            }
            if(value_int!=nodata_int)
            {
              if(first && gridfile!=NULL)
              {
                grid.lon=xllcorner+col*cellsize+cellsize*0.5;
                grid.lat=yllcorner+(nrows-row-1)*cellsize+cellsize*0.5; 
                writefloatcoord(gridfile,&grid);
              }
              data_int[header.ncell][m]=value_int;
              header.ncell++;
            }
          }
      else
        for(row=0;row<nrows;row++)
          for(col=0;col<ncols;col++)
          {
            if(fscanf(file,"%f",&value)!=1)
            {
              fprintf(stderr,"Error reading data at (%d,%d) at year %d and band %d.\n",
                      row+1,col+1,header.firstyear+y,m+1);
              return EXIT_FAILURE;
            }
            if(value!=nodata)
            {
              if(first && gridfile!=NULL)
              {
                grid.lon=xllcorner+col*cellsize+cellsize*0.5;
                grid.lat=yllcorner+(nrows-row-1)*cellsize+cellsize*0.5; 
                writefloatcoord(gridfile,&grid);
              }
              if(type==LPJ_SHORT)
                data[header.ncell][m]=(short)(value*scale);
              else
                data_float[header.ncell][m]=value;
              header.ncell++;
            }
          }
      if(first)
      {
        ncell_first=header.ncell;
        printf("Number of valid grid cells: %d\n",header.ncell);
      }
      else if(header.ncell!=ncell_first)
      {
        fprintf(stderr,"Number of cells %d in '%s' differs from %d.\n",
                header.ncell,argv[y*nbands*nstep+m],ncell_first);
        return EXIT_FAILURE;
      }
      first=FALSE;
      fclose(file);
    }
    switch(type)
    {
      case LPJ_INT:
        for(i=0;i<header.ncell;i++)
          if(fwrite(data_int[i],sizeof(int),nbands*nstep,out)!=nbands*nstep)
          {
            fprintf(stderr,"Error writing output: %s.\n",strerror(errno));
            return EXIT_FAILURE;
          }
        break;
      case LPJ_SHORT:
        for(i=0;i<header.ncell;i++)
          if(fwrite(data[i],sizeof(short),nbands*nstep,out)!=nbands*nstep)
          {
            fprintf(stderr,"Error writing output: %s.\n",strerror(errno));
            return EXIT_FAILURE;
          }
        break;
      case LPJ_FLOAT:
        for(i=0;i<header.ncell;i++)
          if(fwrite(data_float[i],sizeof(float),nbands*nstep,out)!=nbands*nstep)
          {
            fprintf(stderr,"Error writing output: %s.\n",strerror(errno));
            return EXIT_FAILURE;
          }
        break;
      default:
        break;
    }
  }  
  if(gridfile!=NULL)
  {
    /* write header for grid file */
    rewind(gridfile);
    header.nbands=2;
    header.nyear=1;
    header.nstep=1;
    header.timestep=1;
    header.scalar=1;
    header.datatype=LPJ_FLOAT;
    fwriteheader(gridfile,&header,LPJGRID_HEADER,LPJGRID_VERSION);
    fclose(gridfile);
  }
  /* write header for data file */
  rewind(out);
  header.nyear=nyear;
  header.datatype=type;
  header.nbands=nbands;
  header.timestep=1;
  header.nbands=nstep;
  header.scalar=(type==LPJ_SHORT) ? (float)(1./scale) : 1;
  fwriteheader(out,&header,head,version);
  fclose(out);
  if(isjson)
  {
    out_json=malloc(strlen(argv[argc-1])+strlen(JSON_SUFFIX)+1);
    if(out_json==NULL)
    {
      printallocerr("filename");
      return EXIT_FAILURE;
    }
    strcat(strcpy(out_json,argv[argc-1]),JSON_SUFFIX);
    arglist=catstrvec(argv,argc);
    out=fopen(out_json,"w");
    if(out==NULL)
    {
      printfcreateerr(out_json);
      return EXIT_FAILURE;
    }
    fprintjson(out,argv[argc-1],NULL,"asc2clm",NULL,arglist,&header,NULL,NULL,NULL,0,NULL,NULL,NULL,NULL,(gridfile==NULL) ? NULL : &coord_filename,coord_type,CLM,head,FALSE,version);
    fclose(out);
  }
  return EXIT_SUCCESS;
} /* of 'main' */
